### PR TITLE
fix: modify emissions to follow fixed rewards per block

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description
- Replace dynamically calculated block rewards by fixed rewards 
- Add additional logs for mock mainnet deployment 

Closes : https://github.com/zeta-chain/node/issues/1660

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: x/emissions/keeper/block_rewards_components_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+		t.Run(tc.name, func(t *testing.T) {
+			blockRewards, err := emissionskeeper.CalculateFixedValidatorRewards(tc.blockTimeInSecs)
+			assert.NoError(t, err)
+			avgBlockTime, err := sdk.NewDecFromStr(tc.blockTimeInSecs)
+			assert.NoError(t, err)
+			azetaAmountTotalRewards, err := sdk.NewDecFromStr("210000000000000000000000000")
+			numberOfBlocksInAMonth := sdk.NewDec(types.SecsInMonth).Quo(avgBlockTime)
+			numberOfBlocksTotal := numberOfBlocksInAMonth.Mul(sdk.NewDec(12)).Mul(sdk.NewDec(types.EmissionScheduledYears))
+			constantRewardPerBlock := azetaAmountTotalRewards.Quo(numberOfBlocksTotal)
+			assert.Equal(t, constantRewardPerBlock, blockRewards)
```
Comment (lumtis) [inline x/emissions/keeper/block_rewards_components_test.go:48]: We are replicating the logic of the function in the test. This doesn't test the function effectively. We should have the expected output from `CalculateFixedValidatorRewards` in the test case and verify it in the test body

File: x/emissions/keeper/block_rewards_components.go [hunk 2]
```diff
@@ -55,7 +56,26 @@ func (k Keeper) GetDurationFactor(ctx sdk.Context) sdk.Dec {
...
 }
 
-func GetReservesFactor(ctx sdk.Context, keeper types.BankKeeper) sdk.Dec {
-	reserveAmount := keeper.GetBalance(ctx, types.EmissionsModuleAddress, config.BaseDenom)
+func (k Keeper) GetReservesFactor(ctx sdk.Context) sdk.Dec {
+	reserveAmount := k.GetBankKeeper().GetBalance(ctx, types.EmissionsModuleAddress, config.BaseDenom)
 	return sdk.NewDecFromInt(reserveAmount.Amount)
 }
+
+func (k Keeper) GetFixedBlockRewards(ctx sdk.Context) (sdk.Dec, error) {
```
Comment (lumtis) [inline x/emissions/keeper/block_rewards_components.go]: ctx unused

File: x/emissions/abci.go [hunk 3]
```diff
@@ -12,28 +13,47 @@ import (
...
+	}
+	ctx.Logger().Info(fmt.Sprintf("Block Rewards Total:%s Block Height:%d", blockRewards.String(), ctx.BlockHeight()))
 	validatorRewards := sdk.MustNewDecFromStr(keeper.GetParams(ctx).ValidatorEmissionPercentage).Mul(blockRewards).TruncateInt()
 	observerRewards := sdk.MustNewDecFromStr(keeper.GetParams(ctx).ObserverEmissionPercentage).Mul(blockRewards).TruncateInt()
 	tssSignerRewards := sdk.MustNewDecFromStr(keeper.GetParams(ctx).TssSignerEmissionPercentage).Mul(blockRewards).TruncateInt()
-	err := DistributeValidatorRewards(ctx, validatorRewards, keeper.GetBankKeeper(), keeper.GetFeeCollector())
+	ctx.Logger().Info(fmt.Sprintf("Validator Rewards Total:%s , Percentage %s", validatorRewards.String(), keeper.GetParams(ctx).ValidatorEmissionPercentage))
+	ctx.Logger().Info(fmt.Sprintf("Observer Rewards Total:%s , Percentage %s", observerRewards.String(), keeper.GetParams(ctx).ObserverEmissionPercentage))
+	ctx.Logger().Info(fmt.Sprintf("TssSigner Rewards Total:%s , Percentage %s", tssSignerRewards.String(), keeper.GetParams(ctx).TssSignerEmissionPercentage))
+	err = DistributeValidatorRewards(ctx, validatorRewards, keeper.GetBankKeeper(), keeper.GetFeeCollector())
```
Comment (kingpinXD) [inline x/emissions/abci.go:40]: Added additional logs to verify value on mock mainnet

File: common/coin_test.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/zeta-chain/zetacore/common"
+)
+
+func Test_GetAzetaDecFromAmountInZeta(t *testing.T) {
+	tt := []struct {
+		name       string
+		zetaAmount string
+	}{
```
Comment (lumtis) [inline common/coin_test.go:14]: Can we assert the output in these tests as well?

Comment (lumtis) [inline common/coin_test.go:14]: Can we add a test: non valid zetaAmount => error

File: x/emissions/abci.go [hunk 5]
```diff
@@ -12,28 +13,47 @@ import (
 
 func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper) {
 
-	reservesFactor, bondFactor, durationFactor := keeper.GetBlockRewardComponents(ctx)
-	blockRewards := reservesFactor.Mul(bondFactor).Mul(durationFactor)
+	emissonPoolBalance := keeper.GetReservesFactor(ctx)
```
Comment (lumtis) [inline x/emissions/abci.go:16]: Why GetReservesFactor return a balance? Is it the right method used?

File: x/emissions/abci.go [hunk 6]
```diff
@@ -12,28 +13,47 @@ import (
...
+	if emissonPoolBalance.IsZero() {
+		return
+	}
+	blockRewards, err := keeper.GetFixedBlockRewards()
+	if err != nil {
+		ctx.Logger().Error(fmt.Sprintf("Error while getting fixed block rewards %s", err))
+		return
+	}
 	if blockRewards.IsZero() {
+		ctx.Logger().Error("Block rewards are zero")
```
Comment (lumtis) [inline x/emissions/abci.go:26]: Is it an error? I would change it to warn

File: x/emissions/abci.go [hunk 7]
```diff
@@ -12,28 +13,47 @@ import (
...
 		return
 	}
+	if blockRewards.GT(emissonPoolBalance) {
+		ctx.Logger().Error(fmt.Sprintf("Block rewards %s are greater than emission pool balance %s", blockRewards.String(), emissonPoolBalance.String()))
```
Comment (lumtis) [inline x/emissions/abci.go:30]: Same, I think it should be a warn

File: x/emissions/abci.go (hunk 6 above)
Comment (kingpinXD) [reply x/emissions/abci.go:26]: It should be an error , as the rewards are fixed , calculated using constants only

File: x/emissions/abci.go (hunk 5 above)
Comment (kingpinXD) [reply x/emissions/abci.go:16]: ReserversFactor , is the balance of the emission module account , "The amount of token in reserve" . I decided not to rename the function to keep changes minimal

File: x/emissions/abci.go (hunk 7 above)
Comment (kingpinXD) [reply x/emissions/abci.go]: 27c727ade9f75113e34bc452b06286385f41948c

File: common/coin_test.go (hunk 4 above)
Comment (kingpinXD) [reply common/coin_test.go:17]: 27c727ade9f75113e34bc452b06286385f41948c

Comment (kingpinXD) [reply common/coin_test.go:17]: 27c727ade9f75113e34bc452b06286385f41948c

File: x/emissions/abci.go (hunk 5 above)
Comment (brewmaster012) [reply x/emissions/abci.go:16]: typo emisson => emission

**Review by brewmaster012 — APPROVED** [review]

**Review by lumtis — APPROVED** [review]

**Pull request merged**